hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -135,21 +159,122 @@ private IndexResolver() {
+        // Order the index scans based on their collation
+        TreeMap<RelCollation, RelNode> relsTreeMap = new TreeMap<>((coll1, coll2) -> {
+            // Compare the collations field by field
+            int coll1Size = coll1.getFieldCollations().size();
+            int coll2Size = coll2.getFieldCollations().size();
+
+            for (int i = 0; i < coll1Size; ++i) {
+                if (i >= coll2Size) {
+                    // The coll1 has more fields and the prefixes are equal
+                    return 1;
+                }
+
+                RelFieldCollation fieldColl1 = coll1.getFieldCollations().get(i);
+                RelFieldCollation fieldColl2 = coll2.getFieldCollations().get(i);
+                // First, compare directions
+                int cmp = fieldColl1.getDirection().compareTo(fieldColl2.getDirection());
+                if (cmp == 0) {
+                    // Directions are the same
+                    if (fieldColl1.getFieldIndex() == fieldColl2.getFieldIndex()) {
+                        // And fieldIndex is the same, try the next field
+                        continue;
+                    } else {
+                        return Integer.compare(fieldColl1.getFieldIndex(), fieldColl2.getFieldIndex());
+                    }
+                }
+                return cmp;
+            }
+
+            // All the fields from coll1 are equal to the fields from coll2, compare the size
+            return Integer.compare(coll1Size, coll2Size);
+        });
+
+        // Put the rels into the ordered TreeMap
+        for (RelNode rel : rels) {
+            relsTreeMap.put(rel.getTraitSet().getTrait(RelCollationTraitDef.INSTANCE), rel);
+        }
+
+        List<RelNode> resultRels = new ArrayList<>();
+        Map.Entry<RelCollation, RelNode> prevEntry = null;
+        // Go through the ordered collations and exclude covered ones
+        for (Map.Entry<RelCollation, RelNode> entry : relsTreeMap.descendingMap().entrySet()) {
+            RelCollation collation = entry.getKey();
+            RelNode relNode = entry.getValue();
+
+            if (prevEntry == null) {
+                resultRels.add(relNode);
+                prevEntry = entry;
+            } else {
+                RelCollation prevCollation = prevEntry.getKey();
+                if (!coveredCollation(prevCollation, collation)) {
+                    prevEntry = entry;
+                    resultRels.add(relNode);
+                }
+            }
+        }
+        return resultRels;
+    }
+
+    /**
+     * Tests whether collation1 covers (prefix based) collation2
+     *
+     * @param collation1 the collation
+     * @param collation2 the collation
+     * @return {@code true} if collation1 covers collation2, otherwise {@code false}
+     */
+    private static boolean coveredCollation(RelCollation collation1, RelCollation collation2) {
+        if (collation1.getFieldCollations().size() < collation2.getFieldCollations().size()) {
```

Comment:
Perhaps this method could be replaced with `RelCollation.satisfies(RelCollation)`.

Reply:
Indeed, simplified in https://github.com/hazelcast/hazelcast/pull/18013/commits/419012d34cc563494f4437d11e75b315d998e0f1